pallets/dapp-staking-v3/src/types.rs

```diff
@@ -1775,19 +1775,34 @@ impl<MD: Get<u32>, NT: Get<u32>> DAppTierRewards<MD, NT> {
 
     /// Consume reward for the specified dapp id, returning its amount and tier Id.
     /// In case dapp isn't applicable for rewards, or they have already been consumed, returns `None`.
-    pub fn try_claim(&mut self, dapp_id: DAppId) -> Result<(Balance, TierId), DAppTierError> {
+    pub fn try_claim(&mut self, dapp_id: DAppId) -> Result<(Balance, TierAndRank), DAppTierError> {
         // Check if dApp Id exists.
-        let tier_id = self
+        let tier_and_rank = self
             .dapps
             .remove(&dapp_id)
             .ok_or(DAppTierError::NoDAppInTiers)?;
 
-        Ok((
-            self.rewards
-                .get(tier_id as usize)
-                .map_or(Balance::zero(), |x| *x),
-            tier_id,
-        ))
+        let (tier_id, rank) = tier_and_rank.destruct();
+
+        let mut amount = self
+            .rewards
+            .get(tier_id as usize)
+            .map_or(Balance::zero(), |x| *x);
+
+        if !rank.is_zero() {
+            // TODO: make sure this is the correct approach & add tests
```

Comment:
@Dinonard this implements the logic on issue #1198
There's one problem with this approach. If dapp makes enough to qualify for tier 1 but because tier 1 is full it will end up into tier 2. So because it has made enough to qualify as rank 10 and we distribute rewards like this then this dapp will receive reward equal to tier 1.

Reply:
one approach will be to configure the maximum reward per rank. let's say delta reward between tier 1 and 2 is 500K - 400K = 100K then max additional reward will be (half) 50K, so tier2 rank 10 will receive max 450K

Reply:
The algorithm/logic within `fn get_dapp_tier_assignment_and_rewards` should ensure tier & ranking assignment respect the reward boundaries though.

I don't believe it should change much compared to what we have now, except as discussed, instead of just observing tier capacity as number of slots, we should also account for tier reward pools.

E.g. scenario:
* 10 ranks between tiers
* Tier1 has capacity of 3 slots
* dApps **A** and **B**  meet the requirements to enter tier 1
* dApp **C** meets the requirement to enter tier 2, and also to enter rank 7
* dApp **D** meets the requirement to enter tier 2, and also to enter rank 6
* dApp **E** meets the requirement to enter tier 2, and also to enter rank 2

In such scenario, since it's a competition, what should happen:
* dApps **A** and **B** enter Tier1
* dApp **C** enters tier 2, and has ranking 7, earning rewards in respect to that rank
* dApp **D** enters tier 2, but there's no enough reward pool capacity to support rank 6 so it gets downgraded to rank 3
* dApp **E** enters tier 2, but there's no remaining reward pool capacity to support rank 2, so it gets degraded to rank 0 (pure tier 2 reward)

What do you think?

Reply:
yeah that can work

Reply:
@Dinonard There's still an issue. What if dapp C,D,E also qualify for tier 1 but there's no space, so D,E ends on tier 2 with rank 10 and if we use the proposed solution dapps D,E will make reward the same as tier 1 dapps.

Reply:
what if D and E with tier 2 and rank 10 make twice the reward. `amount = reward_per_slot + rank * reward_per_slot / 10`?

Reply:
I might be misunderstanding the question, but I don't see that as a problem since it's already _solved_ in the solution we have live right now.

The current solution is that if we have more dApps qualify for a tier than there's capacity, only the top ones will get into that tier. The others will be pushed down to lower tiers, if there are any.

A basic example would be a scenario where only 1 tier exists, and it has capacity for 10 dApps. If 11 dApps pass the entry threshold, only the top 10 will enter and earn rewards, while the 11th dApp won't earn anything.

The amount staked on a dApp is merely a guiding line in which tier it ends up, it's still a competition. Theoretically, even if you have high enough nomination to enter tier 1, you might end up e.g. in tier 3 since compared to others, they just have higher nomination.

---

To get back to your example, in case C, D, E qualify for tier 1 but we only have capacity for 3 dApps:
* Let's assume C's stake is higher than D's stake, and D's stake is higher than E's stake (for simplicity)
* C enters tier 1, earning full tier 1 reward
* D enters tier 2, rank 0, earning just tier 2 reward (it cannot have any ranking because reward pool has been exhausted)
* E enters tier 2, rank 0, same reasoning as dApp D

Reply:
yes but D has a lot of stake (qualify for tier1) but there's no slot so ends up on tier 2 with rank 10

Reply:
It cannot end up with rank 10 if there's no capacity for rewards.

Total amount staked on dAPp tells the max possible **tier & rank** assignment it can get.
Remaining reward pool capacity tells what's the actual **tier & rank** it will be getting.

Since after dApps A, B and C assignment to tier 1 the entire reward pool for that tier has been exhausted, dApp D can only be assigned tier 2 with rank 0.

It's the same approach as it's taken now, only the limiting factor isn't strictly number of slots, but amount of rewards.

Reply:
yes that's what I have implemented so far on this draft. It also makes sure the reward covers the rank

Reply:
Great! Will review when ready! 🙂 

From the discussion we've had here, I'd suggest we make a descriptive forum post update once solution is stable enough.
And definitely a good description in the docs.